### PR TITLE
add types and interfaces

### PR DESCRIPTION
**Pull request by bhavya3024**
Add KiteConnect Types and Interfaces according to the best of my knowledge
cc @vividvilla @ranjanrak.

### CODE REVIEW

**Comment by bhavya3024** [comment]
@vividvilla @ranjanrak I only am facing one issue, while running the `npm run build-test` command, it compiles all the typescript files and adds javacscript files along with it. I tried so much, but unable to avoid.
@anurag-roy can you see this issue as well ?

File: constants/index.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    'market.quote.ltp': '/quote/ltp',
+
+    'gtt.triggers': '/gtt/triggers',
+    'gtt.trigger_info': '/gtt/triggers/{trigger_id}',
+    'gtt.place': '/gtt/triggers',
+    'gtt.modify': '/gtt/triggers/{trigger_id}',
+    'gtt.delete': '/gtt/triggers/{trigger_id}'
+};
+
+export const DEFAULTS: AnyObject = {
```
Comment (vividvilla) [inline constants/index.ts:56]: I think these default settings should be moved to constants.

File: interfaces/connect.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    timeout?: number;
+    /*
+     * Default url for which user will be redirected to login page
+     * defa
+     */
+}
+
+export enum ExchangeTypes {
+    NSE = 'NSE',
+    BSE = 'BpSE',
```
Comment (vividvilla) [inline interfaces/connect.ts:123]: There is a typo here.

File: interfaces/connect.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    NFO = 'NFO',
+    BFO = 'BFO',
+    CDS = 'CDS',
+    MCX = 'MCX'
+};
+
+export enum TransactionTypesCaps {
```
Comment (vividvilla) [inline interfaces/connect.ts:130]: you don't need to suffix these types with `caps` .

File: interfaces/connect.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    MARGIN_COMMODITY = 'commodity',
+}
+
+export enum StatusTypes {
+    STATUS_CANCELLED = 'CANCELLED',
+    STATUS_REJECTED = 'REJECTED',
+    STATUS_COMPLETE = 'COMPLETE',
+}
+
+export enum GitStatusTypes {
```
Comment (vividvilla) [inline interfaces/connect.ts:59]: Typo, should be `GTTStatusTypes`

**Review by vividvilla — COMMENTED** [review]
Added few comments, haven't reviewed all the files yet.

File: constants/index.ts (hunk 1 above)
Comment (bhavya3024) [reply constants/index.ts:56]: these are already inside `constants/index.ts`

**Comment by bhavya3024** [comment]
@vividvilla @ranjanrak Resolved most of the comments, but I am unable to avoid js files getting created along with ts files while running tsc for test files. But any ways I put all of the js files into .gitignore!

**Comment by anurag-roy** [comment]
@bhavya3024
Run `tsc` to generate the .js files, run the tests and finally run
```
tsc --build --clean
```
to get rid of the generated .js files

File: interfaces/connect.ts [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    SLM = 'SL-M',
+    MARKET = 'MARKET'
+};
+
+export enum Validities {
+    DAY = 'DAY',
+    IOC = 'IOC',
+};
+
+export interface OrderParams {
```
Comment (anurag-roy) [inline interfaces/connect.ts:158]: Every property is not optional, please remove `?` from mandatory properties.

File: interfaces/connect.ts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    sip_id?: string | number;
+    trigger_type?: GTTStatusTypes.GTT_TYPE_OCO | GTTStatusTypes.GTT_TYPE_SINGLE;
+    trigger_values?: any[];
+    last_price?: string | number;
+    orders?: Order[];
+    order_id?: string | number;
+    parent_order_id?: string;
+}
+
+export interface ModifyOrderParams {
```
Comment (anurag-roy) [inline interfaces/connect.ts:216]: Same as above.

File: interfaces/connect.ts [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    order_id?: string;
+};
+
+export interface CancelOrderParams { 
+    parent_order_id?: string;
+    variety?: string;
+    order_id?: string; 
+};
+
+export interface Order {
```
Comment (anurag-roy) [inline interfaces/connect.ts:187]: Same as above.

File: interfaces/connect.ts [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    validity_ttl?: number;
+    iceberg_legs?: number;
+    iceberg_quantity?: number;
+    auction_number?: number;
+    tag?: string;
+    variety?: string;
+    order_id?: string;
+};
+
+export interface CancelOrderParams {
```
Comment (anurag-roy) [inline interfaces/connect.ts:181]: Same as above.

File: interfaces/connect.ts [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+};
+
+export enum OrderTypes {
+    ORDER_TYPE_MARKET = 'MARKET',
+    ORDER_TYPE_LIMIT = 'LIMIT',
+    ORDER_TYPE_SL = 'SL',
+    ORDER_TYPE_SLM = 'SLM',
+};
+
+export enum Varities {
```
Comment (anurag-roy) [inline interfaces/connect.ts:18]: Typo

File: interfaces/connect.ts (hunk 5 above)
Comment (bhavya3024) [reply interfaces/connect.ts:158]: can you tell which one is mandatory. I intetionally did them to get rid of errors!

Comment (anurag-roy) [reply interfaces/connect.ts:158]: You can refer to jsdoc comments in the master branch (Optional properties had an `optional` attribute) or go to the current [kiteconnect docs](https://kite.trade/docs/kiteconnectjs/v3/index.html) and there you can see all the optional properties are marked optional, rest are mandatory.

e.g.
image

File: lib/connect.ts [hunk 10]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @memberOf KiteConnect
-     * @instance
-     * @param {string} variety Order variety (ex. bo, co, amo)
-     * @param {string} order_id ID of the order.
-     * @param {Object} [params] Order params.
-regular).
-     * @param {string} [params.parent_order_id] Parent order id incase of multilegged orders.
-     */
-    self.cancelOrder = function (variety, order_id, params) {
+    cancelOrder(variety: Varieties | string, order_id: string | number, params?: Order) {
```
Comment (anurag-roy) [inline lib/connect.ts:277]: variety should be of type `Varieties` only

File: lib/connect.ts [hunk 11]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @param {string} orders[].order_type   Order type (MARKET, LIMIT etc.)
-     * @param {number} orders[].quantity   Quantity of the order
-     * @param {number} orders[].price   Price at which the order is going to be placed (LIMIT orders)
-     * @param {number} orders[].trigger_price   Trigger price (for SL, SL-M, CO orders)
-     * @param {string} mode    (optional) Compact mode will only give the total margins
-     */
-    self.orderMargins = function(orders, mode=null) {
-        return _post("order.margins", orders, null, null, true,
-                    {"mode":mode});
+    orderMargins(orders: Order[], mode = null) {
```
Comment (anurag-roy) [inline lib/connect.ts:304]: `mode` can have the value `'compact'`

File: interfaces/connect.ts [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+     * Defaults to `null`
+     */
+    access_token?: string | null;
+    /**
+     * API end point root. Unless you explicitly want to send API requests to a
+     * non-default endpoint, this can be ignored.
+     *
+     * Defaults to 'https?://api.kite.trade'
+     */
+    root?: string | RegExp | Url;
```
Comment (anurag-roy) [inline interfaces/connect.ts:96]: Type for root is only `string`

File: lib/connect.ts [hunk 13]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-    /**
-     * Get list of order history.
-     * @method getOrderHistory
-     * @memberOf KiteConnect
-     * @instance
-     * @param {string} order_id ID of the order whose order details to be retrieved.
-     */
-    self.getOrderHistory = function(order_id) {
-        return  _get("order.info", {"order_id": order_id}, null, formatResponse);
+    getOrderHistory(order_id: string | number) {
```
[the remark this answers is not in the document]
Comment (bhavya3024) [reply lib/connect.ts:292]: in tests, it is given a number as well

File: lib/connect.ts (hunk 11 above)
Comment (bhavya3024) [reply lib/connect.ts:304]: null is the default value, compct may come from the client!.

**Comment by bhavya3024** [comment]
@anurag-roy @vividvilla @ranjanrak 
Resolved most of the comments, also I had to create more interfaces to specify params for each function that which arguments are required and optional. Also needed to add more arguments and types in params w.r.t unit test cases.

File: lib/connect.ts (hunk 11 above)
Comment (anurag-roy) [reply lib/connect.ts:304]: So shouldn't the type be something like this:
```ts
orderMargins(orders: Order[], mode: 'compact' | null = null)
```

**Comment by anurag-roy** [comment]
@bhavya3024 I'll try to complete my review by tonight. Also, we need to add return types/interfaces for all the functions.

**Comment by bhavya3024** [comment]
@anurag-roy added types in return types of functions as well as per my best of my knowledge
cc @vividvilla @ranjanrak

**Comment by bhavya3024** [comment]
@vividvilla @ranjanrak!

File: lib/connect.ts [hunk 14]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-        } else if (error.request) {
-            // The request was made but no response was received
-            // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
-            // http.ClientRequest in node.js
-            resp.error_type = "NetworkException";
-            resp.message = "No response from server with error code: " + error.code;
-        } else if (error.message) {
-            resp = error
-        }
+        }, function (error) {
```
Comment (ranjanrak) [inline lib/connect.ts:151]: Missed specifying the type of error, either you can define `(error: Error)` as the desired type. If you want to allow any type for the error parameter, you can use `(error: any)`

File: lib/connect.ts [hunk 15]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-    /**
-     * Invalidate the refresh token.
-     * @method invalidateRefreshToken
-     * @memberOf KiteConnect
-     * @instance
-     * @param {string} refresh_token Token to invalidate.
-     */
-    self.invalidateRefreshToken = function(refresh_token) {
-        return _delete("api.token.invalidate", {
+    invalidateRefreshToken = function (refresh_token: string): Promise<AxiosResponse<any>> {
```
Comment (ranjanrak) [inline lib/connect.ts:247]: Maybe you can follow the same function signature here as well like the rest of all (renewAccessToken, etc) to keep it constant. Suggests maintaining consistency in function signatures across your codebase.
```
invalidateRefreshToken(refresh_token:string):Promise<AxiosResponse<any>> 
{...}
```

File: lib/connect.ts [hunk 16]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @param {string} orders[].order_type   Order type (MARKET, LIMIT etc.)
-     * @param {number} orders[].quantity   Quantity of the order
-     * @param {number} orders[].price   Price at which the order is going to be placed (LIMIT orders)
-     * @param {number} orders[].trigger_price   Trigger price (for SL, SL-M, CO orders)
-     * @param {string} mode    (optional) Compact mode will only give the total margins
-     */
-    self.orderMargins = function(orders, mode=null) {
-        return _post("order.margins", orders, null, null, true,
-                    {"mode":mode});
+    orderMargins(orders: Order[], mode = null): Promise<AxiosResponse<any>> {
```
Comment (ranjanrak) [inline lib/connect.ts:304]: Missing types for orders and mode. Maybe use: `orderMargins(orders:Record<string, any>, mode:string)`

File: lib/connect.ts [hunk 17]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @param {number} orders[].quantity   Quantity of the order
-     * @param {number} orders[].price   Price at which the order is going to be placed (LIMIT orders)
-     * @param {number} orders[].trigger_price   Trigger price (for SL, SL-M, CO orders)
-     * @param {string} consider_positions    Boolean to consider users positions while calculating margins
-     * @param {string} mode    (optional) Compact mode will only give the total margins
-     */
-    self.orderBasketMargins = function(orders, consider_positions=true, mode=null) {
-        return _post("order.margins.basket", orders, null, null, true,
-                    {"consider_positions":consider_positions, "mode":mode});
+    orderBasketMargins(orders: Order[], consider_positions = true, mode = null): Promise<AxiosResponse<any>> {
```
Comment (ranjanrak) [inline lib/connect.ts:309]: Missing types for orders, consider_positions and mode. Suggestion:
`orderBasketMargins(orders:Record<string, any>, consider_positions:boolean=true, mode:string="")`

File: lib/connect.ts [hunk 18]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @memberOf KiteConnect
-     * @instance
-     * @param {Array} [segment] Filter instruments based on exchange (NSE, BSE, NFO, BFO, CDS, MCX).
-     * If no `segment` is specified, all instruments are returned.
-     */
-    self.getInstruments = function(exchange) {
-        if(exchange) {
-            return _get("market.instruments", {
-                "exchange": exchange
+    getInstruments(exchange: boolean): Promise<AxiosResponse<any>> {
```
Comment (ranjanrak) [inline lib/connect.ts:330]: The incorrect exchange type is being used; it should be a `string` instead of a `boolean`.

File: lib/connect.ts [hunk 19]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-     * @method getTriggerRange
-     * @memberOf KiteConnect
-     * @instance
-     * @param {string} exchange Exchange in which instrument is listed (NSE, BSE, NFO, BFO, CDS, MCX).
-     * @param {string} tradingsymbol Tranding symbol of the instrument (ex. RELIANCE, INFY).
-     * @param {string} transaction_type Transaction type (BUY or SELL).
-     */
-    self.getTriggerRange = function(transaction_type, instruments) {
-        return _get("market.trigger_range",
+    getTriggerRange(transaction_type: TransactionTypes, instruments: string[]): Promise<AxiosResponse<any>> {
```
Comment (ranjanrak) [inline lib/connect.ts:368]: getTriggerRange function is already depreciated. It can be removed.

File: lib/connect.ts [hunk 20]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-    function _post(route, params, responseType?, responseTransformer?, isJSON = false, queryParams: any =null) {
-        return request(route, "POST", params || {}, responseType, responseTransformer, isJSON, queryParams);
-    }
+    return data;
+}
 
-    function _put(route, params, responseType?, responseTransformer?, isJSON = false, queryParams=null) {
-        return request(route, "PUT", params || {}, responseType, responseTransformer, isJSON, queryParams);
-    }
+function parseHistorical(jsonData: AnyObject): AnyObject {
```
Comment (ranjanrak) [inline lib/connect.ts:637]: Better to make `parseHistorical` function private `private parseHistorical(jsonData: AnyObject): AnyObject`.

File: lib/connect.ts [hunk 21]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
+}
 
-        // Replace variables in "RESTful" URLs with corresponding params
-        if(uri.indexOf("{") !== -1) {
-            let k;
-            for(k in params) {
-                if(params.hasOwnProperty(k)) {
-                    uri = uri.replace("{" + k + "}", params[k]);
-                }
+function transformInstrumentsResponse(data: any, headers: AnyObject): unknown[] | any {
```
Comment (ranjanrak) [inline lib/connect.ts:664]: It is advisable to designate the `transformInstrumentsResponse` function as a private function.

File: lib/connect.ts [hunk 22]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
-            } else {
-                // post url encoded payload
-                payload = querystring.stringify(params);
+        return parsedData;
+    }
+
+    return data;
+}
+
+function transformMFInstrumentsResponse(data: any, headers: AnyObject): any {
```
Comment (ranjanrak) [inline lib/connect.ts:685]: It is advisable to designate the `transformMFInstrumentsResponse` function as a private function.

File: lib/connect.ts [hunk 23]
```diff
@@ -65,1354 +69,648 @@ import utils from "./utils";
...
+            }
         }
+    }
 
-        return data;
+    return data;
+}
+
+// Format response ex. datetime string to date
+function formatResponse(data: AnyObject): AnyObject {
```
Comment (ranjanrak) [inline lib/connect.ts:606]: It is advisable to designate the `formatResponse` function as a private function.

**Review by ranjanrak — COMMENTED** [review]
Hi @bhavya3024 
I have left a couple of reviews. You can check them. Additionally, doc comments are missing. I would like to hear your thoughts on incorporating [TSDoc standardized comments notation](https://tsdoc.org/pages/tags/alpha/) here.

**Comment by GargDivanshu** [comment]
hey @bhavya3024 , @ranjanrak told me in a separate comment thread, you are working on types and interface additions. Can you let me know from where and how I can start helping you with this?

**Comment by bhavya3024** [comment]
@ranjanrak I have fixed the PR comments.
For comments in TypeDoc notataion, I have used a vscode extension to auto generate comments for `typescript` files and used typedoc to generate `docs`

**Comment by bhavya3024** [comment]
@ranjanrak @vividvilla

**Comment by bhavya3024** [comment]
@ranjanrak @vividvilla https://github.com/zerodha/kiteconnectjs/pull/96#issuecomment-1581165526, any updates on this ? I have already added the docs using TypeDoc

**Comment by ranjanrak** [comment]
LGTM. Thanks @bhavya3024 for this PR! 👍 I just have a few minor changes in mind, I'll take care of those. Thanks again for your contribution!

**Comment by bhavya3024** [comment]
> hey @bhavya3024 , @ranjanrak told me in a separate comment thread, you are working on types and interface additions. Can you let me know from where and how I can start helping you with this?

@GargDivanshu its done now!

**Pull request merged**